File: 5. LinkedLists/6. zipper_lists/zipper_lists.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

a = Node("a")
b = Node("b")
c = Node("c")
a.next = b
b.next = c
# a -> b -> c

x = Node("x")
y = Node("y")
z = Node("z")
x.next = y
y.next = z
# x -> y -> z

def print_list(head):
  output = ""
  while(head.next is not None):
    output += str(head.val) + "->"
    head = head.next
  output += str(head.val)
  print(output)

# Iterative
def zipper_lists(head_1, head_2):
  if(head_1 is None and head_2 is None):
    return None
  if(head_1 is None):
    return head_2
  if(head_2 is None):
    return head_1
  zipperHead = head_1

  while(head_1 is not None and head_2 is not None):
    nextHead1 = head_1.next
    nextHead2 = head_2.next
    head_1.next = head_2
    if nextHead1 is not None:
      head_2.next = nextHead1
    head_1 = nextHead1
    head_2 = nextHead2
  return zipperHead

# Iterative 2
def zipper_lists1(head_1, head_2):
  tail = head_1
  current_1 = head_1.next
  current_2 = head_2
  count = 0
  while current_1 is not None and current_2 is not None:
    if count % 2 == 0:
      tail.next = current_2
      current_2 = current_2.next
    else:
      tail.next = current_1
      current_1 = current_1.next
    tail = tail.next
    count += 1
    
  if current_1 is not None:
    tail.next = current_1
  if current_2 is not None:
    tail.next = current_2
    
  return head_1
    
# Recursive 
def zipper_lists2(head_1, head_2):
  if(head_1 is None and head_2 is None):
    return None
  if(head_1 is None):
    return head_2
  if(head_2 is None):
    return head_1
  
  nextHead1 = head_1.next
  head_1.next = head_2
  head_2.next = zipper_lists2(nextHead1,head_2.next)
  
  return head_1

print_list(a)
print_list(x)
print_list(zipper_lists(a, x))
# print_list(zipper_lists1(a, x))
# print_list(zipper_lists1(a, x))
# a -> x -> b -> y -> c -> z

File: 5. LinkedLists/6. zipper_lists/zipper_lists.js

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// a.next = b;
// b.next = c;
// // a -> b -> c

// const x = new Node("x");
// const y = new Node("y");
// const z = new Node("z");
// x.next = y;
// y.next = z;
// // x -> y -> z

// Iterative
const zipperLists = (head1, head2) => {
    if(head1 === null && head2 === null) return null;
    if(head1 === null) return head2;
    if(head2 === null) return head1;
    let zipperHead = head1;
    while(head1 !== null && head2 !== null){
        let next1 = head1.next;
        let next2 = head2.next;
        head1.next = head2;
        if(next1)
            head2.next = next1;
        head1 = next1;
        head2 = next2;
    }
    return zipperHead;
    
}

// Iterative - structy
// const zipperLists = (head1, head2) => {
//     let tail = head1;
//     let head = head1;
//     let current1 = head1.next;
//     let current2 = head2;
//     let count = 0;
//     while(current1 !== null && current2 !== null){
//         if((count % 2) === 0){
//             tail.next = current2;
//             current2 = current2.next;
//         }else{
//             tail.next = current1;
//             current1 = current1.next;
//         }
//         tail = tail.next;
//         count++;
//     }
//     if(current1 !== null) tail.next = current1;
//     if(current2 !== null) tail.next = current2;

//     return head;
// }

// Recursive
// const zipperLists = (head1, head2) => {
//     if(head1 === null && head2 === null) return null;
//     if(head1 === null) return head2;
//     if(head2 === null) return head1;
//     let next1 = head1.next;
//     head1.next = head2;
//     head2.next = zipperLists(next1, head2.next);
//     return head1;
// }

// const linkedListValues = (head, arr=[]) => {
//     if(head === null) return [];
//     arr = [head.val, ...linkedListValues(head.next)]
//     return arr;
// }

// let arr1 = linkedListValues(a);
// console.log(`Initial linked list 1: ${arr1}`);

// let arr2 = linkedListValues(x);
// console.log(`Initial linked list 2: ${arr2}`);

// let newHead = zipperLists(a, x);

// let arr3 = linkedListValues(newHead);
// console.log(`Zipped List: ${arr3}`);


// Complexity:
// Iterative:
// n = length of list 1
// m = length of list 2
// Time: O(min(n, m))
// Space: O(1)

// Recursive:
// n = length of list 1
// m = length of list 2
// Time: O(min(n, m))
// Space: O(1)


